### ISSUE TITLE
1.19.2 kube-apiserver can't write to etcd - "transport is closing"

### ISSUE THREAD

**Issue by der-eismann** [kubernetes/kubernetes#95334]
**What happened**: We run a self-hosted kubernetes v1.18.2 cluster on Flatcar Container Linux and are testing an upgrade to v1.19.2. While doing this on a separate staging cluster we noticed that the kube-apiserver has problems writing to etcd (meaning etcd is empty) without any obvious reason.

**What you expected to happen**: kube-apiserver running as smoothly as always without any errors.

**How to reproduce it (as minimally and precisely as possible)**: Setup a new kubernetes v1.19.2 cluster with etcd 3.3.12

**Anything else we need to know?**: These are the logs from the kube-apiserver:
```
I1006 17:05:18.242080       1 client.go:360] parsed scheme: "passthrough"
I1006 17:05:18.242122       1 passthrough.go:48] ccResolverWrapper: sending update to cc: {[{http://0.etcd.k8s.domain.loc:2379  <nil> 0 <nil>}] <nil> <nil>}
I1006 17:05:18.242249       1 clientconn.go:948] ClientConn switching balancer to "pick_first"
I1006 17:05:18.242587       1 balancer_conn_wrappers.go:78] pickfirstBalancer: HandleSubConnStateChange: 0xc00c6da8b0, {CONNECTING <nil>}
I1006 17:05:18.245726       1 balancer_conn_wrappers.go:78] pickfirstBalancer: HandleSubConnStateChange: 0xc00c6da8b0, {READY <nil>}
I1006 17:05:18.246748       1 controlbuf.go:508] transport: loopyWriter.run returning. connection error: desc = "transport is closing"
I1006 17:05:22.842730       1 client.go:360] parsed scheme: "passthrough"
I1006 17:05:22.842767       1 passthrough.go:48] ccResolverWrapper: sending update to cc: {[{http://1.etcd.k8s.domain.loc:2379  <nil> 0 <nil>}] <nil> <nil>}
I1006 17:05:22.842776       1 clientconn.go:948] ClientConn switching balancer to "pick_first"
I1006 17:05:22.842889       1 balancer_conn_wrappers.go:78] pickfirstBalancer: HandleSubConnStateChange: 0xc00c81ee70, {CONNECTING <nil>}
I1006 17:05:22.844120       1 balancer_conn_wrappers.go:78] pickfirstBalancer: HandleSubConnStateChange: 0xc00c81ee70, {READY <nil>}
I1006 17:05:22.844902       1 controlbuf.go:508] transport: loopyWriter.run returning. connection error: desc = "transport is closing"
I1006 17:05:31.308740       1 client.go:360] parsed scheme: "passthrough"
I1006 17:05:31.308814       1 passthrough.go:48] ccResolverWrapper: sending update to cc: {[{http://2.etcd.k8s.domain.loc:2379  <nil> 0 <nil>}] <nil> <nil>}
I1006 17:05:31.308825       1 clientconn.go:948] ClientConn switching balancer to "pick_first"
I1006 17:05:31.308956       1 balancer_conn_wrappers.go:78] pickfirstBalancer: HandleSubConnStateChange: 0xc00337c010, {CONNECTING <nil>}
I1006 17:05:31.311857       1 balancer_conn_wrappers.go:78] pickfirstBalancer: HandleSubConnStateChange: 0xc00337c010, {READY <nil>}
I1006 17:05:31.313169       1 controlbuf.go:508] transport: loopyWriter.run returning. connection error: desc = "transport is closing"
```

However the etcd cluster is healthy and I can read and write from a separate etcd pod inside the cluster:
```
$ etcdctl cluster-health
member c7549a1f541ef0dc is healthy: got healthy result from http://1.etcd.k8s.domain.loc:2379
member e0a9942f6b765c5f is healthy: got healthy result from http://0.etcd.k8s.domain.loc:2379
member f795bfa714149988 is healthy: got healthy result from http://2.etcd.k8s.domain.loc:2379
cluster is healthy
```

I don't think that the etcd version is playing a huge role since someone has [reported the same issue](https://github.com/kubernetes/kubernetes/issues/94155#issuecomment-689331049) (comment in #94155) with k8s v1.19.0 and etcd 3.4.13.

**Environment**:
- Kubernetes version: v1.19.2
- Cloud provider or hardware configuration: AWS
- OS: Flatcar Container Linux 2605.6.0
- Kernel: 5.4.66-flatcar
- Install tools: Manually provisioned with Rancher kubelet container & official `kube-*` images
- Network plugin and version (if this is a network-related bug): Flannel v0.11.0
- Others:
    - Etcd: 3.3.12

**Comment by caesarxuchao** [comment]
cc @cheftako

**Comment by cheftako** [comment]
cc @nckturner

**Comment by jingyih** [comment]
> we noticed that the kube-apiserver has problems writing to etcd (meaning etcd is empty) without any obvious reason.

Could you elaborate on this? I am trying to understand if the messages you posted (e.g. `transport: loopyWriter.run returning. connection error: desc = "transport is closing"`) are just noisy logs or do they actually causing writes to fail.

Could you try `kubectl get --raw='/readyz?verbose'` and look for the entry for etcd? Internally the kube apiserver will try to perform a quorum read to etcd servers as part of the health checking.

**Comment by der-eismann** [comment]
Thanks for your answer!

> > we noticed that the kube-apiserver has problems writing to etcd (meaning etcd is empty) without any obvious reason.
> 
> Could you elaborate on this? I am trying to understand if the messages you posted (e.g. `transport: loopyWriter.run returning. connection error: desc = "transport is closing"`) are just noisy logs or do they actually causing writes to fail.

Sorry, this was a wrong assumption. On our production cluster I was able to view existing resources with
```
$ etcdctl ls /registry/statefulsets/default
```
This was missing on the new staging cluster, however I just noticed that this is the old v2 API and the keys are unused. If I use
```
$ ETCDCTL_API=3 etcdctl get --prefix=true ""
```
I am able to view all existing resources, so the etcd connection seems to be fine.

> Could you try `kubectl get --raw='/readyz?verbose'` and look for the entry for etcd? Internally the kube apiserver will try to perform a quorum read to etcd servers as part of the health checking.

```
$ kubectl get --raw='/readyz?verbose'
[+]ping ok
[+]log ok
[+]etcd ok
[+]informer-sync ok
[+]poststarthook/start-kube-apiserver-admission-initializer ok
[+]poststarthook/generic-apiserver-start-informers ok
[+]poststarthook/start-apiextensions-informers ok
[+]poststarthook/start-apiextensions-controllers ok
[+]poststarthook/crd-informer-synced ok
[+]poststarthook/bootstrap-controller ok
[+]poststarthook/rbac/bootstrap-roles ok
[+]poststarthook/scheduling/bootstrap-system-priority-classes ok
[+]poststarthook/start-cluster-authentication-info-controller ok
[+]poststarthook/aggregator-reload-proxy-client-cert ok
[+]poststarthook/start-kube-aggregator-informers ok
[+]poststarthook/apiservice-registration-controller ok
[+]poststarthook/apiservice-status-available-controller ok
[+]poststarthook/kube-apiserver-autoregistration ok
[+]autoregister-completion ok
[+]poststarthook/apiservice-openapi-controller ok
[+]shutdown ok
healthz check passed
```

Looks like the connection is fine and this is just noisy (& confusing) logging.

**Comment by ialidzhikov** [comment]
@der-eismann , can we then close the issue?

**Comment by der-eismann** [comment]
I'd prefer it if the issue was fixed instead of just closing it. Even if there is no actual problem these entries are clogging the logs and something must cause them.
I just tried it with 1.19.3, the issue still persists.

**Comment by ialidzhikov** [comment]
@der-eismann, so what is the issue - the noisy and confusing logging? If that is the case, then please kindly open a new issue (there are too much comments on this issue and it is hard to follow what was the initial problem and what are your complaints currently). From the currently issue title and description, it looks like 1.19.2 apiserver cannot write to etcd at all which is not the case as you also confirmed.

**Comment by KeithTt** [comment]
Same problem here, but there is really problem with this node, not noise. Some pods which scheduled to this node can not run normally.

Kubernetes version: v1.16.4
Docker version: 18.09.9

And, everything seems healthy.
```shell
# kubectl get --raw='/readyz?verbose'
[+]ping ok
[+]log ok
[+]etcd ok
[+]poststarthook/generic-apiserver-start-informers ok
[+]poststarthook/start-apiextensions-informers ok
[+]poststarthook/start-apiextensions-controllers ok
[+]poststarthook/crd-informer-synced ok
[+]poststarthook/bootstrap-controller ok
[+]poststarthook/rbac/bootstrap-roles ok
[+]poststarthook/scheduling/bootstrap-system-priority-classes ok
[+]poststarthook/ca-registration ok
[+]poststarthook/start-kube-apiserver-admission-initializer ok
[+]poststarthook/start-kube-aggregator-informers ok
[+]poststarthook/apiservice-registration-controller ok
[+]poststarthook/apiservice-status-available-controller ok
[+]poststarthook/kube-apiserver-autoregistration ok
[+]autoregister-completion ok
[+]poststarthook/apiservice-openapi-controller ok
[+]shutdown ok
healthz check passed
```

The docker logs show unmount message:
```text
Jul 26 10:44:16 m-69-224 dockerd[2159]: time="2021-07-26T10:44:16.502695136+08:00" level=warning msg="299f8fed9ea212d1d2682b457ae081c6f615314889cc2fcbcd375eb1ad05ac34 cleanup: failed to unmount IPC: umount /var/lib/docker/containers/299f8fed9ea212d1d2682b457ae081c6f615314889cc2fcbcd375eb1ad05ac34/mounts/shm, flags: 0x2: no such file or directory"
Jul 26 10:44:19 m-69-224 dockerd[2159]: time="2021-07-26T10:44:19.547344743+08:00" level=info msg="ignoring event" module=libcontainerd namespace=moby topic=/tasks/delete type="*events.TaskDelete"
Jul 26 10:48:19 m-69-224 dockerd[2159]: time="2021-07-26T10:48:19.812878576+08:00" level=info msg="ignoring event" module=libcontainerd namespace=moby topic=/tasks/delete type="*events.TaskDelete"
```

But I do not find the container in `/var/lib/docker/containers/`...
